### ISSUE TITLE
Migrate out of unsafe / deprecated React lifecycle methods

### ISSUE THREAD

**Issue by WholeMilk** [palantir/blueprint#2384]
Under React 16.3+ with [React.StrictMode](https://reactjs.org/docs/strict-mode.html), several components are producing warnings in regards to unsafe React lifecycle methods usage as they will be [deprecated in future React 16.x release and removed in React 17.0](https://reactjs.org/blog/2018/03/27/update-on-async-rendering.html#gradual-migration-path).

**Comment by giladgray** [comment]
yes we've gotta do this! happy to accept PRs to convert components one-by-one.

**Comment by llorca** [comment]
oof actually this won't happen any time soon as we still support React 15.

**Comment by phpnode** [comment]
Do you happen to know when support for React 15 will end? It is possible to polyfill the new component methods for React 15 using https://github.com/reactjs/react-lifecycles-compat if you're happy to go with that solution I'll open a PR.

**Comment by demigor** [comment]
React Concurrent Rendering (16.7) will not be supported as long as old lifecycle methods have been used in Blueprint. Please, raise the prio of this issue.

**Comment by adidahiya** [comment]
@phpnode thanks for the link, react-lifecycles-compat looks promising. I think we would like to support React 15 for a few more months, so backwards compatibility using a shim layer would be appropriate.

**Comment by emlynmac** [comment]
A year has passed since this came up and we're approaching some react changes. As of today:

```
Warning: componentWillReceiveProps has been renamed, and is not recommended for use. See https://fb.me/react-async-component-lifecycle-hooks for details.

* Move data fetching code or side effects to componentDidUpdate.
* If you're updating state whenever props change, refactor your code to use memoization techniques or move it to static getDerivedStateFromProps. Learn more at: https://fb.me/react-derived-state
* Rename componentWillReceiveProps to UNSAFE_componentWillReceiveProps to suppress this warning in non-strict mode. In React 17.x, only the UNSAFE_ name will work. To rename all deprecated lifecycles to their new names, you can run `npx react-codemod rename-unsafe-lifecycles` in your project source folder.

Please update the following components: Blueprint3.Spinner,  Blueprint3.OverflowList, Blueprint3.Overlay, Blueprint3.Popover, Blueprint3.QueryList, Blueprint3.Suggest, Blueprint3.Toaster, Blueprint3.Dialog, Blueprint3.Alert
```

**Comment by adidahiya** [comment]
fixed by #3702, but we are still dealing with lingering semantic issues after the migration
